target/sh.c

```diff
@@ -58,32 +68,59 @@ static void sh_emit_inst(Inst* inst) {
     break;
 
   case ADD:
-    emit_line("%s=$(( ($%s + %s) & " UINT_MAX_STR " ))",
+    emit_line(": $(( %s = (($%s + %s) & " UINT_MAX_STR ") ))",
               reg_names[inst->dst.reg],
               reg_names[inst->dst.reg], sh_src_str(inst));
     break;
 
   case SUB:
-    emit_line("%s=$(( ($%s - %s) & " UINT_MAX_STR "))",
+    emit_line(": $(( %s = (($%s - %s) & " UINT_MAX_STR ") ))",
               reg_names[inst->dst.reg],
               reg_names[inst->dst.reg], sh_src_str(inst));
     break;
 
   case LOAD:
-    emit_line("eval z=\\$m%s", sh_src_str(inst));
-    emit_line("%s=$(( $z + 0 ))", reg_names[inst->dst.reg]);
+    emit_line(": $(( %s = m%s + 0 ))", reg_names[inst->dst.reg], sh_src_str(inst));
     break;
 
   case STORE:
-    emit_line("eval m%s=$%s", sh_src_str(inst), reg_names[inst->dst.reg]);
+    emit_line(": $(( m%s = $%s ))", sh_src_str(inst), reg_names[inst->dst.reg]);
     break;
 
   case PUTC:
-    emit_line("t=$((%s&255))", sh_src_str(inst));
-    emit_line("printf \"\\\\$(printf '%%03o' $t)\"");
+    emit_line(": $(( t = (%s & 255) ))", sh_src_str(inst));
+    emit_line("printf \\\\$((t/64))$((t/8%%8))$((t%%8))");
     break;
 
   case GETC:
+#ifdef POSIX_SHELL
+    // The POSIX standard doesn't support read -n1, so we read a full line and buffer it.
```

Comment:
It seems bash's `read -n1` immediately returns after the first character is typed (not when the line ends). As a result, the portable implementation isn't 100% equivalent to `read -n1` so I'm keeping the existing implementation for users that want to write interactive scripts.